[LBSE] Take transforms into account on mask/clip content

rwlbuis opened the pull request:
#### 2b69c8b0d1bf5df630b1e099a0e9af04fef6353f
<pre>
[LBSE] Take transforms into account on mask/clip content
<a href="https://bugs.webkit.org/show_bug.cgi?id=267075">https://bugs.webkit.org/show_bug.cgi?id=267075</a>

Reviewed by Nikolas Zimmermann.

Take transforms into account on mask/clip content by adding logic to calculateClipContentRepaintRect
and calculateMaskContentRepaintRect, which is used by LBSE only.

Fix another clip problem related to transforms since for &lt;use&gt; computeClipPath did not take
the transform on the referenced element into account.

Also, clipper uses masks once it is determined masks are needed, and does not switch to clips for referenced clipPaths.

Finally, when resource bounding box for masks is fully outside of the object bounding box )no intersection), use
object bounding box rect as fallback.

* LayoutTests/platform/ios/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/platform/ios/svg/repaint/mask-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/W3C-SVG-1.1/masking-mask-01-b-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/W3C-SVG-1.1/masking-path-01-b-expected.png:
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::computeClipPath const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::calculateClipContentRepaintRect):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::calculateMaskContentRepaintRect):

Canonical link: <a href="https://commits.webkit.org/273046@main">https://commits.webkit.org/273046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/866f385c4ca4a2cfc8c3eda470172b873ceb1129

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36064 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36727 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30887 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35163 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15270 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10023 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29929 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34591 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10901 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30411 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9619 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30442 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38028 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30759 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35714 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9742 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/7655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33618 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11528 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7846 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10308 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10555 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
```diff
@@ -285,6 +286,11 @@ Path RenderSVGModelObject::computeClipPath(AffineTransform& transform) const
 
+    if (auto* useElement = dynamicDowncast<SVGUseElement>(element())) {
+        if (auto* clipChild = useElement->rendererClipChild())
+            transform.multiply(downcast<RenderLayerModelObject>(*clipChild).checkedLayer()->currentTransform(RenderStyle::individualTransformOperations()).toAffineTransform());
```
Great solution! I should have suggested that before 😂

nikolaszimmermann reviewed (approved):
Great! r=me

rwlbuis committed:
[LBSE] Take transforms into account on mask/clip content
https://bugs.webkit.org/show_bug.cgi?id=267075

Reviewed by Nikolas Zimmermann.

Take transforms into account on mask/clip content by adding logic to calculateClipContentRepaintRect
and calculateMaskContentRepaintRect, which is used by LBSE only.

Fix another clip problem related to transforms since for <use> computeClipPath did not take
the transform on the referenced element into account.

Also, clipper uses masks once it is determined masks are needed, and does not switch to clips for referenced clipPaths.

Finally, when resource bounding box for masks is fully outside of the object bounding box )no intersection), use
object bounding box rect as fallback.

* LayoutTests/platform/ios/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/platform/ios/svg/repaint/mask-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/W3C-SVG-1.1/masking-mask-01-b-expected.png:
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/W3C-SVG-1.1/masking-path-01-b-expected.png:
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::computeClipPath const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::calculateClipContentRepaintRect):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::calculateMaskContentRepaintRect):

Canonical link: https://commits.webkit.org/273046@main
